File: app/src/main/java/com/example/bindechexclock/ui/romans/RomansFragment.java
package com.example.bindechexclock.ui.romans;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider; // Changed
import androidx.annotation.Nullable; // Added for onViewCreated
import android.os.Bundle; // Added for onViewCreated

import com.example.bindechexclock.MainActivity;
import com.example.bindechexclock.R;
import com.example.bindechexclock.TimeManager;

public class RomansFragment extends Fragment {

    private RomansViewModel romansViewModel; // Changed to private

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        romansViewModel =
                new ViewModelProvider(this).get(RomansViewModel.class); // Changed
        View root = inflater.inflate(R.layout.fragment_romans, container, false);
        final TextView textView = root.findViewById(R.id.text_romans);
        romansViewModel.getText().observe(getViewLifecycleOwner(), new Observer<String>() {
            @Override
            public void onChanged(@Nullable String s) {
                textView.setText(s);
            }
        });
        return root;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getActivity() instanceof MainActivity) {
            MainActivity mainActivity = (MainActivity) getActivity();
            TimeManager timeManager = mainActivity.getTimeManager();

            timeManager.getRomanTimeLiveData().observe(getViewLifecycleOwner(), new Observer<String>() {
                @Override
                public void onChanged(String romanTimeValue) {
                    String prefix = getString(R.string.time_in) + " " + getString(R.string.title_romans) + "\n";
                    romansViewModel.setmText(prefix + romanTimeValue);
                }
            });
        }
    }
}

File: app/src/main/java/com/example/bindechexclock/ui/romans/RomansFragment.kt
package com.example.bindechexclock.ui.romans

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels // Required for by viewModels()
import androidx.lifecycle.Observer
import com.example.bindechexclock.MainActivity
import com.example.bindechexclock.R
import com.example.bindechexclock.TimeManager

class RomansFragment : Fragment() {

    // Use the 'by viewModels()' Kotlin property delegate from fragment-ktx
    private val romansViewModel: RomansViewModel by viewModels()
    private var textViewRomans: TextView? = null // To hold the TextView

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_romans, container, false)
        textViewRomans = root.findViewById(R.id.text_romans)

        // Observe LiveData from RomansViewModel
        romansViewModel.text.observe(viewLifecycleOwner, Observer { s ->
            textViewRomans?.text = s
        })
        return root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Access TimeManager via MainActivity
        (activity as? MainActivity)?.let { mainActivity ->
            val timeManager: TimeManager = mainActivity.timeManager // Accessing Kotlin property

            // Observe LiveData from TimeManager
            timeManager.romanTimeLiveData.observe(viewLifecycleOwner, Observer { romanTimeValue ->
                // getString requires a Context, which Fragment provides
                val prefix = getString(R.string.time_in) + " " + getString(R.string.title_romans) + "\n"
                romansViewModel.updateText(prefix + romanTimeValue)
            })
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        // Clear the reference to the view to prevent memory leaks
        textViewRomans = null
    }
}
